Add the yk BlockDisambiguate pass.

vext01 opened the pull request:
This is required for yk to unambiguously map machine blocks back to IRblocks.

See the doc comment in [`BlockDisambiguate.cpp`](https://github.com/ykjit/ykllvm/pull/11/files#diff-6fe45f14428841c1072fb97dfa17cc765f4de37342460c80ace7125111a40d9eR1) for a detailed description of the problem this solves and how we solve it.

[that comment took many hours to get right -- it's really "detaily" and hard to explain]

A companion PR will come soon. This PR should be merged first.

ltratt commented:
To make sure that I understand: we split the LLVM *IR* so that the code gen backend doesn't feel the need to introduce internal control flow?

ltratt commented:
I think I've got it. Maybe silly question: should we fix the code gen so that it never inserts additional control flow or is that impractical / silly / ... ?

vext01 commented:
> should we fix the code gen so that it never inserts additional control flow or is that impractical / silly / ... ?

I doubt it will be possible. I few months back I looked at what kinds of operations do this, but I've since forgot.

I think some LLVM IR instructions are high-level enough that they have to lower to extra control flow.

Also finding all of the places where this happens would be hard and maybe platform specific.

ltratt commented:
> I think some LLVM IR instructions are high-level enough that they have to lower to extra control flow.

Right, I think I see where you're going with this.

ltratt commented on llvm/lib/Transforms/Yk/BlockDisambiguate.cpp:
```diff
@@ -0,0 +1,204 @@
+//===- BlockDisambiguate.cpp - Unambiguous block mapping for yk ----===//
+//
+// This pass ensures that yk is able to unambiguously map machine blocks back
```
This comment is very long (which is good!) but reads like a murder mystery (which is bad). Let's give the reader an intuition up-front about what they'll see. Maybe something like "LLVM IR blocks that branch to themselves lead to multiple machine-level basic blocks that we cannot unambiguously map back to LLVM IR. This pass splits such blocks into two parts such that the resulting machine-level basic blocks do have an unambiguous mapping back to the IR."

Something I also haven't quite got from the comment is the precise preconditions for when this pass will decide to rewrite things. It's probably worth clarifying that somewhere (not necesarilly up-front though).

vext01 replied:
Can I squash those last three commits to make the review easier for you?

ltratt replied:
Go ahead.

vext01 commented:
106b29c reworks the comment.

I've added a up-front spoiler to the whodunnit, simplified/shortened some bits, made the text sound more like an informal proof, and added an "alternative approaches" section.

Hopefully this is more coherent?

ltratt commented:
OK, that helps me a lot -- thanks!

I think I have one last question. Is the problem here that we're trying to deal with someone jumping to a machine basic block that is not the first machine basic block corresponding to an LLVM IR basic block?

In other words, imagine I have an LLVM IR basic block `L1` and from it are generated two machine basic blocks `M1` and `M2`. Can anyone ever jump to `M2` without having executed `M1`? [If "yes", I agree with the PR as-is. If "no" then I wonder if an alternative solution is to never try and map something that's not the first machine basic block (i.e. in this case `M1`).]

vext01 commented:
> Is the problem here that we're trying to deal with someone jumping to a machine basic block that is not the first machine basic block corresponding to an LLVM IR basic block?

No. That's not the problem I'm tying to address here. I don't think it can happen.

> If "no" then I wonder if an alternative solution is to never try and map something that's not the first machine basic block (i.e. in this case M1).]

There are two related problematic cases related to the fact that calls in LLVM IR don't terminate a block:

 - Tracing can be turned on mid-IR-block (in a machine block other than the first). If we didn't map this machine block, then we could miss a chunk of program semantics.
 - We require an entry for when a callee returns to a caller.

To expand on the latter, suppose we have ir like:
```
main() {
  ...stuff
  call f()
  ...more stuff
}
```

Currently the mapper requires a sequence like:
```
main:bb0   <- execute "stuff" up until the call to f
f:bb0
...
f:bbN
main:bb0    <- return to main to execute "more stuff"
```

The second `main:bb0` is required to tell the trace compiler that it should resume copying instructions from after the call to `f`. The trace compiler has to keep track of how far it has gotten through each active IR block on the stack.

[Lukas' block-splitting branch (that we are still unsure if we will merge) aims to simplify this stuff by ensuring that each IR block has at most one call. Then there's no need to keep track of "progress through blocks"]

vext01 commented:
> Would that obviate the need for this current PR then?

Sadly I don't think so.

The example in the big doc-comment has no calls, so Lukas' pass would leave it untouched, and the issue would remain.

ltratt commented:
> There are two related problematic cases related to the fact that calls in LLVM IR don't terminate a block:
>
>  *  Tracing can be turned on mid-IR-block (in a machine block other than the first). If we didn't map this machine block, then we could miss a chunk of program semantics.
>  *  We require an entry for when a callee returns to a caller.

Devil's advocate: Lukas's patch means that the second case above would no longer be a problem (I think? Meaning we need only handle the first case. But could we say "we throw away the first block in a trace if there's not a precise match"?

vext01 commented:
>Lukas's patch means that the second case above would no longer be a problem (I think? Meaning we need only handle the first case.

That's correct.

>  But could we say "we throw away the first block in a trace if there's not a precise match"?

Not sure I understand. Can you give an example?

ltratt commented:
> >  But could we say "we throw away the first block in a trace if there's not a precise match"?
>Not sure I understand. Can you give an example?

This relates to:

> Tracing can be turned on mid-IR-block (in a machine block other than the first). If we didn't map this machine block, then we could miss a chunk of program semantics.

We could (for example) insert a dummy loop after turning PT on, so that we guarantee to record all the bits of the program we care about?

vext01 commented:
Thinking about this some more (and my head is on the verge of exploding), you can only turn on tracing via a call, namely to the control point. So Lukas' pass would in fact help in the case that tracing is turned on mid-IR-block. Since the code after the call to the control point would be in a separate block.

I'm still unsure about what you mean with "throw away the first block in a trace if there's not a precise match" though.

I'm starting to get confused.

ltratt commented:
OK, so imagine you have this:

```
BB1
start_tracing()
BB2
BB3
```

and PT kicks in half-way through BB2. This PR, I think, guarantees that we pick up BB2. What I'm saying is that if PT kicks in half-way through BB2, we should drop that from the trace i.e. we would see it starting at BB3. If we care about BB2, we should insert a "fake" basic block before BB2.

vext01 commented:
Are those blocks high-level IR blocks, or machine blocks. And is this after Lukas' pass has run?

ltratt commented:
Those are machine basic blocks and after Lukas's PR has merged.

ltratt commented:
I'm more-or-less fine with this -- except I can't see there being either a) a test which this new pass changes b) a new test which this pass does something specific on. Should I be expecting one of those?

vext01 commented:
> a) a test which this new pass changes

On the `yk` side, [several tests were strengthened](https://github.com/ykjit/yk/pull/434/files). These would fail prior to the new pass.

ltratt commented:
I think we're ready to squash!

vext01 commented:
Great news.

Let review that one small change [on the yk side](https://github.com/ykjit/yk/pull/434) first, then i will push a squash.

vext01 commented:
squashed.

ltratt commented:
bors r+
